### ISSUE TITLE
RFC - JS-style promises vs the current promise-like multithreading system.

### ISSUE THREAD

**Issue by j3pic** [j3pic/cl-async-await#1]
**TL;DR:** Dear users: Should I revert to the pre-rewrite version of this library that implemented JS-like promises (with `then` and `catch` methods, and some macros to implement the `async` and `await` stuff), or should I carry on with the current system?

After commit https://github.com/j3pic/cl-async-await/commit/aeef11cb6b08d47fe9cfb1fc46b7191d8766aecb, I rewrote the library. Before the rewrite, the functionality provided was analogous to what is provided in JavaScript: A `promise` was an object that contained a closure that would be called with two other closures giving the options to "resolve" or "reject" the promise. There was a `then` generic function and a `catch` generic function that handled those two events. If a condition was signalled, that was translated into a call to the "reject" closure.

In addition to this, there were macros to create "async functions" that implicitly returned promise objects instead of evaluating their bodies immediately. There was an `await-let*` macro that was analogous to JavaScript's `await` keyword. The form expanded to a series of calls to the `then` generic function on the values of the variables being bound. The value of an `await-let*` form was a promise object. A subset of the current library's functionality could be obtained by using the `parallel-promise` class. What was missing was the ability to invoke a restart on a rejected promise.

The current version of CL-ASYNC-AWAIT is really a threading library. "promise objects" are really threads whose values are returned across a message queue. In some ways, these promise objects behave identically to JavaScript-style promise objects, without the need to invert the program into continuation-passing style via macros. Because there's no need to invert the program's structure, `await` is simply a generic function that reads the return value from the promise's message queue and engages in an error-handling protocol if an error is passed. 

There are a few major disadvantages to this approach:

1. It doesn't fit well with callback-oriented GUI libraries. Using the old version of the library, your callback functions can be closures over promises, and you could write a GUI program in a style that looks procedural. This can be worked around, but you end up with a bunch of threads that do nothing but wait for messages from GUI callbacks and forward them to the caller of the `await` function.
2. It's especially ill-suited for COCOA, which, being poorly designed, requires most operations to be invoked from the primary thread. In the old version of CL-ASYNC-AWAIT, this would pose no problem, and neither would the fact that COCOA is allergic to exception handling.
3. The current version of CL-ASYNC-AWAIT does little more than duplicate a subset of the functionality of the LPARALLEL library. 

If you're a user of this library, please leave a comment indicating whether you'd like to see this library reverted to commit https://github.com/j3pic/cl-async-await/commit/aeef11cb6b08d47fe9cfb1fc46b7191d8766aecb, which is the last commit before the rewrite, or if you think the current implementation is better.
